web/src/lib/actions/shortcut.svelte.ts

```diff
@@ -0,0 +1,315 @@
+import {
+  alwaysTrueFactory,
+  type KeyCombo,
+  keyDownListenerFactory,
+  type KeyDownListenerFactory,
+  type KeyInput,
+  normalizeKeyInput,
+  type ShortcutCallback,
+} from '$lib/actions/input';
+import ShortcutsModal from '$lib/modals/ShortcutsModal.svelte';
+import { modalManager } from '@immich/ui';
+import { untrack } from 'svelte';
+import { t } from 'svelte-i18n';
+import type { Attachment } from 'svelte/attachments';
+import { SvelteMap } from 'svelte/reactivity';
+import { get } from 'svelte/store';
+
+export enum Category {
+  Application = 'app_actions',
+  AssetActions = 'asset_actions',
+  ViewActions = 'view_actions',
+  QuickActions = 'quick_actions',
+  Navigation = 'navigation',
+  Selection = 'selection',
+}
+
+export const getCategoryString = (category: Category) => get(t)(category);
+
+export const category = (category: Category, text: string, variant?: ShortcutVariant): ShortcutHelp => {
+  return {
+    variant,
+    category,
+    text,
+  };
+};
+
+const explicitCategoryList = [
+  Category.QuickActions,
+  Category.AssetActions,
+  Category.ViewActions,
+  Category.Selection,
+  Category.Navigation,
+  Category.Application,
+];
+
+export const sortCategories = (categories: Category[]) =>
+  [...categories].sort((a, b) => {
+    const indexA = explicitCategoryList.indexOf(a);
+    const indexB = explicitCategoryList.indexOf(b);
+    return (indexA === -1 ? Infinity : indexA) - (indexB === -1 ? Infinity : indexB);
+  });
+
+export enum ShortcutVariant {
+  SelectAll,
+  DeselectAll,
+  AddAlbum,
+  AddSharedAlbum,
+  PrevAsset,
+  NextAsset,
+  Delete,
+  PermDelete,
+  PreviousAsset,
+  PreviousDay,
+  NextDay,
+  PreviousMonth,
+  NextMonth,
+  PreviousYear,
+  NextYear,
+  Trash,
+  Search,
+  SearchFilter,
+  FocusNext,
+  FocusPrevious,
+}
+
+type ShortcutHelp = {
+  variant?: ShortcutVariant;
+  category?: Category;
+  text: string;
+  info?: string;
+};
+export type KeyboardHelp = ShortcutHelp & { key: string[][] };
+
+type InternalKeyboardHelp = KeyboardHelp & { scope: number; $InternalHelpId: string };
+type KeyTargets = HTMLElement | Document | Window;
+
+const isMacOS = /Mac(intosh|Intel)/.test(globalThis.navigator.userAgent);
+
+// state variables
+let helpArray: InternalKeyboardHelp[] = $state([]);
+// eslint-disable-next-line svelte/no-unnecessary-state-wrap
+let shortcutVariants = $state(new SvelteMap<ShortcutVariant, ShortcutVariant>());
+let currentScope = $state(0);
+let showingShortcuts = $state(false);
+
+const activeScopeShortcuts: KeyboardHelp[] = $derived(
+  helpArray.filter((helpObjectArrayObject) => helpObjectArrayObject.scope === currentScope),
+);
+
+function isLetter(c: string) {
+  return c.toLowerCase() != c.toUpperCase();
+}
+
+const expandKeys = (shortcuts: KeyCombo[]) => {
+  return shortcuts.map((s) => {
+    const keys: string[] = [];
```

Comment:
Semantically speaking I personally preferred the way we had it before that the modifiers had their dedicated fields in an object and could be `true/false`